pkg/openevec/adam.go

```diff
@@ -25,3 +31,155 @@ func (openEVEC *OpenEVEC) AdamStart() error {
 	}
 	return nil
 }
+
+// UploadSigningCert uploads the provided signing certificate to the OpenEVEC controller.
+func (openEVEC *OpenEVEC) UploadSigningCert(newSignCert []byte) error {
+	changer := &adamChanger{}
+	ctrl, dev, err := changer.getControllerAndDevFromConfig(openEVEC.cfg)
+	if err != nil {
+		return fmt.Errorf("getControllerAndDevFromConfig: %w", err)
+	}
+
+	// we need to re-encrypt existing configs with the new certificate because EVE has support only for one server signing certificate
+	err = reencryptConfigs(ctrl, dev, newSignCert)
+	if err != nil {
+		return fmt.Errorf("failed to reencrypt existing configs: %w", err)
+	}
+
+	if err = changer.setControllerAndDev(ctrl, dev); err != nil {
+		return fmt.Errorf("setControllerAndDev: %w", err)
+	}
+
+	err = ctrl.UploadSigningCert(newSignCert)
+	if err != nil {
+		log.Errorf("cannot upload signing cert: %s", err.Error())
+	} else {
+		log.Infof("Signing cert uploaded")
+	}
+	return nil
+}
+
+func reencryptConfigs(ctrl controller.Cloud, dev *device.Ctx, newSignCert []byte) error {
+	// get device certificate from the controller
+	devCert, err := ctrl.GetECDHCert(dev.GetID())
+	if err != nil {
+		return fmt.Errorf("cannot get device certificate from cloud: %w", err)
+	}
+
+	// get signing certificate from the controller
+	oldSignCert, err := ctrl.SigningCertGet()
+	if err != nil {
+		log.Error("cannot get cloud's signing certificate. will use plaintext")
+		return nil
+	}
+
+	edenHome, err := utils.DefaultEdenDir()
+	if err != nil {
+		return fmt.Errorf("DefaultEdenDir: %w", err)
+	}
+	keyPath := filepath.Join(edenHome, defaults.DefaultCertsDist, "signing-key.pem")
+	ctrlPrivKey, err := os.ReadFile(keyPath)
+	if err != nil {
+		return fmt.Errorf("cannot read %s: %w", keyPath, err)
+	}
+
+	oldCryptoConfig, err := utils.GetCommonCryptoConfig(devCert, oldSignCert, ctrlPrivKey)
+	if err != nil {
+		return fmt.Errorf("GetCommonCryptoConfig: %w", err)
+	}
+
+	newCryptoConfig, err := utils.GetCommonCryptoConfig(devCert, newSignCert, ctrlPrivKey)
+	if err != nil {
+		return fmt.Errorf("GetCommonCryptoConfig: %w", err)
+	}
+
+	cipherCtx, err := utils.CreateCipherCtx(newCryptoConfig)
+	if err != nil {
+		return fmt.Errorf("CreateCipherCtx: %w", err)
+	}
+	utils.AddCipherCtxToDev(dev, cipherCtx)
+
+	// re-encrypt all app configs with the new signing certificate
+	appConfigs := ctrl.ListApplicationInstanceConfig()
+	for _, config := range appConfigs {
+		if config.CipherData != nil {
+			encBlock, err := utils.CryptoConfigUnwrapper(config.CipherData, oldCryptoConfig)
+			if err != nil {
+				return fmt.Errorf("CryptoConfigUnwrapper error: %w", err)
+			}
+
+			cipherData, err := utils.CryptoConfigWrapper(encBlock, newCryptoConfig, cipherCtx)
+			if err != nil {
+				return fmt.Errorf("CryptoConfigWrapper error: %w", err)
+			}
+
+			// set cipher data encrypted with the new signing certificate
+			config.CipherData = cipherData
+		}
```

Comment:
this code block is repeated for 4 different types of config here - they all have a `CipherData` field - but to put it in a function for reuse I would need to create an interface like the following in the eve-api repo. do you think it's suitable for that?
```go
type CipherDataHolder interface {
	GetCipherData() []byte
	SetCipherData(data []byte)
}
```

Reply:
Yes, it would make sense to avoid repetition. But you can simply define the interface here in eden. Go code in eve-api repo is generated from proto and not meant for manual editing.

Reply:
True, but I would need to define the setter methods on those `config` structs and I can only do that local to the original definition

Reply:
I see. Maybe some [reflect](https://pkg.go.dev/reflect) magic can do the job?

Reply:
Btw. wouldn't Get method (which is available) enough? You can change the value of existing CipherData rather than to create a new instance and changing the pointer value.

Reply:
You are right, I changed the value, but still had to jump through some hoops to avoid copying the lock that's part of every protobuf message